File: python/models/task.py
from datetime import date, datetime, time, timezone

DUE_DATE_KEY = "dueDate"
OVERDUE_KEY = "overdue"
SORT_BY_DUE_DATE = "dueDate"
SORT_OPTIONS = (SORT_BY_DUE_DATE,)

_UTC_DESIGNATOR = "Z"
_UTC_OFFSET = "+00:00"
_DATE_ONLY_LENGTH = 10  # "YYYY-MM-DD" — a value of any other length carries a time
_END_OF_DAY = time.max
_UNDATED_PLACEHOLDER = datetime.min.replace(tzinfo=timezone.utc)

_tasks = []
_next_id = 1


def _now_utc():
    """
    Purpose:
    Supplies the current instant used for every overdue comparison.

    Input:
    None

    Output:
    Timezone-aware datetime in UTC
    """
    return datetime.now(timezone.utc)


def _with_utc_offset(value):
    """
    Purpose:
    Rewrites a trailing UTC designator into the numeric offset this interpreter accepts.

    Input:
    ISO 8601 string

    Output:
    Equivalent ISO 8601 string that `datetime.fromisoformat` can parse
    """
    if value.endswith(_UTC_DESIGNATOR):
        return value[: -len(_UTC_DESIGNATOR)] + _UTC_OFFSET
    return value


def parse_due_date(value):
    """
    Purpose:
    Normalizes a client-supplied due date into the single representation the store holds.
    A date without a time means the end of that day in server-local time.

    Input:
    None, or an ISO 8601 date or datetime string

    Output:
    Canonical UTC ISO 8601 string, or None when no due date was supplied.
    Raises ValueError for any value that is not a usable ISO 8601 date or datetime
    """
    if value is None:
        return None
    if not isinstance(value, str):
        raise ValueError("due date must be a string")
    try:
        if len(value) == _DATE_ONLY_LENGTH:
            moment = datetime.combine(date.fromisoformat(value), _END_OF_DAY)
        else:
            moment = datetime.fromisoformat(_with_utc_offset(value))
        if moment.tzinfo is None:
            moment = moment.astimezone()
        return moment.astimezone(timezone.utc).isoformat()
    except (TypeError, OverflowError) as exc:
        raise ValueError("due date is out of the supported range") from exc


def _parse_stored_due_date(value):
    """
    Purpose:
    Turns a stored due date back into an instant for comparison and ordering.

    Input:
    Stored due date string, or None

    Output:
    Timezone-aware datetime, or None when there is no usable due date
    """
    if value is None:
        return None
    try:
        moment = datetime.fromisoformat(_with_utc_offset(value))
    except (ValueError, TypeError):
        # Normalization on write makes this unreachable; a bad stored value must not
        # fail the whole read.
        return None
    if moment.tzinfo is None:
        return moment.replace(tzinfo=timezone.utc)
    return moment


def _is_overdue(task, due, now):
    """
    Purpose:
    Decides whether a task is late: it has a due date, that date is strictly past,
    and the task is not completed.

    Input:
    Stored task dict, its parsed due date or None, current instant in UTC

    Output:
    Boolean
    """
    if task["completed"] or due is None:
        return False
    return due < now


def _due_date_sort_key(entry):
    """
    Purpose:
    Orders tasks by due date ascending, tasks without one last, ties broken by id.

    Input:
    Tuple of parsed due date, task id, and task view

    Output:
    Tuple usable as a `sorted` key
    """
    due, task_id, _ = entry
    return (due is None, due or _UNDATED_PLACEHOLDER, task_id)


def _view(task, due, now):
    """
    Purpose:
    Copies a stored task and attaches the derived overdue flag, leaving the store untouched.

    Input:
    Stored task dict, its parsed due date or None, current instant in UTC

    Output:
    New dict with every stored field plus `overdue`
    """
    view = dict(task)
    view[OVERDUE_KEY] = _is_overdue(task, due, now)
    return view


def task_view(task):
    """
    Purpose:
    Builds the API representation of one task.

    Input:
    Stored task dict

    Output:
    New dict with every stored field plus `overdue`
    """
    due = _parse_stored_due_date(task.get(DUE_DATE_KEY))
    return _view(task, due, _now_utc())


def task_list_view(sort=None):
    """
    Purpose:
    Builds the API representation of the whole list, optionally ordered by due date,
    without ever reordering or mutating the stored list.

    Input:
    Sort key, or None for insertion order

    Output:
    New list of task view dicts
    """
    now = _now_utc()
    entries = []
    for task in _tasks:
        due = _parse_stored_due_date(task.get(DUE_DATE_KEY))
        entries.append((due, task["id"], _view(task, due, now)))
    if sort == SORT_BY_DUE_DATE:
        entries = sorted(entries, key=_due_date_sort_key)
    return [view for _, _, view in entries]


def create_task(title, due_date=None):
    """
    Purpose:
    Appends a new task to the store.

    Input:
    Title string, optional already-normalized due date string

    Output:
    The stored task dict
    """
    global _next_id
    task = {
        "id": _next_id,
        "title": title,
        "completed": False,
        DUE_DATE_KEY: due_date,
    }
    _next_id += 1
    _tasks.append(task)
    return task


def get_tasks():
    return _tasks


def get_task(task_id):
    return next((t for t in _tasks if t["id"] == task_id), None)


def update_task(task_id, completed):
    task = get_task(task_id)
    if task is None:
        return None
    task["completed"] = completed
    return task


def delete_task(task_id):
    task = get_task(task_id)
    if task is None:
        return False
    _tasks.remove(task)
    return True


def reset():
    global _next_id
    _tasks.clear()
    _next_id = 1

File: Node/models/task.js
let tasks = [];
let nextId = 1;

// Default timezone for due-date timestamps.
const IST_OFFSET = '+05:30';

// Normalize an incoming due date to a canonical ISO-8601 timestamp in IST.
// - null/undefined stays null.
// - A date-only value (YYYY-MM-DD) defaults to start of day (00:00) in IST.
// - A datetime without an offset is interpreted as IST.
// - A datetime with an offset is converted to the equivalent IST wall-clock.
function normalizeDueDate(value) {
  if (value === null || value === undefined) return null;
  const s = String(value).trim();

  if (/^\d{4}-\d{2}-\d{2}$/.test(s)) {
    return `${s}T00:00:00${IST_OFFSET}`;
  }

  const m = /^(\d{4}-\d{2}-\d{2})T(\d{2}):(\d{2})(?::(\d{2}))?$/.exec(s);
  if (m) {
    const [, date, hh, mm, ss] = m;
    return `${date}T${hh}:${mm}:${ss || '00'}${IST_OFFSET}`;
  }

  // Value carries an offset (or Z): convert to the equivalent IST wall-clock.
  const d = new Date(s);
  const ist = new Date(d.getTime() + (5 * 60 + 30) * 60000);
  const pad = (n) => String(n).padStart(2, '0');
  return (
    `${ist.getUTCFullYear()}-${pad(ist.getUTCMonth() + 1)}-${pad(ist.getUTCDate())}` +
    `T${pad(ist.getUTCHours())}:${pad(ist.getUTCMinutes())}:${pad(ist.getUTCSeconds())}${IST_OFFSET}`
  );
}

function createTask(title, dueDate = null) {
  const task = { id: nextId++, title, completed: false, due_date: normalizeDueDate(dueDate) };
  tasks.push(task);
  return task;
}

function getTasks() {
  // Ascending by due_date timestamp (earliest first); null due dates last,
  // preserving insertion order within the null group. Array.sort is stable.
  return [...tasks].sort((a, b) => {
    const aNull = a.due_date === null || a.due_date === undefined;
    const bNull = b.due_date === null || b.due_date === undefined;
    if (aNull && bNull) return 0;
    if (aNull) return 1;
    if (bNull) return -1;
    return new Date(a.due_date).getTime() - new Date(b.due_date).getTime();
  });
}

function getTask(id) {
  return tasks.find((t) => t.id === id);
}

function updateTask(id, updates) {
  const task = getTask(id);
  if (!task) return null;
  Object.assign(task, updates);
  return task;
}

function setDueDate(id, dueDate) {
  const task = getTask(id);
  if (!task) return null;
  task.due_date = normalizeDueDate(dueDate);
  return task;
}

function deleteTask(id) {
  const index = tasks.findIndex((t) => t.id === id);
  if (index === -1) return false;
  tasks.splice(index, 1);
  return true;
}

function reset() {
  tasks = [];
  nextId = 1;
}

module.exports = { createTask, getTasks, getTask, updateTask, setDueDate, deleteTask, reset };
